core/Rename.cpp

```diff
@@ -184,6 +195,72 @@ namespace olympia
     void Rename::handleFlush_(const FlushManager::FlushingCriteria & criteria)
     {
         ILOG("Got a flush call for " << criteria);
+        // Restore the rename map, reference counters and freelist by walking through the inst_queue
+        while (inst_queue_.size())
+        {
+            const auto & inst_ptr = inst_queue_.back();
+            if (!criteria.includedInFlush(inst_ptr))
+            {
+                break;
+            }
+            else
+            {
+                auto const & dests = inst_ptr->getDestOpInfoList();
+                for (const auto & dest : dests)
+                {
+                    // restore rename table following a flush
+                    const auto rf  = olympia::coreutils::determineRegisterFile(dest);
+                    const auto num = dest.field_value;
+                    const bool is_x0 = (num == 0 && rf == core_types::RF_INTEGER);
+                    if (!is_x0)
+                    {
+                        auto const & original_dest = inst_ptr->getRenameData().getOriginalDestination();
+                        map_table_[rf][num] = original_dest.val;
+
+                        // free renamed PRF mapping when reference counter reaches zero
+                        auto const & renamed_dest = inst_ptr->getRenameData().getDestination();
+                        --reference_counter_[renamed_dest.rf][renamed_dest.val];
+                        if(reference_counter_[renamed_dest.rf][renamed_dest.val] <= 0){
+                            freelist_[renamed_dest.rf].push(renamed_dest.val);
+                        }
+                    }
+                    else
+                    {
```

Comment:
We probably can remove this empty else right? If an instruction destination is x0, we don't need to rewind anything in terms of flushing.

Reply:
Good catch, I must have missed it in the merge - removed.